e2e/test/scenarios/joins/joins-reproductions.cy.spec.js

```diff
@@ -1262,4 +1262,28 @@ describe("issue 42385", { tags: "@external" }, () => {
       cy.findByLabelText("Right column").should("not.exist");
     });
   });
+
+  it("should remove invalid join clause in incomplete draft state when query database changes (metabase#42385)", () => {
+    openOrdersTable({ mode: "notebook" });
+    join();
+    entityPickerModal().within(() => {
+      entityPickerModalTab("Tables").click();
+      cy.findByText("Products").click();
+    });
+
+    getNotebookStep("join")
+      .findByLabelText("Right table")
+      .findByText("Products")
+      .click();
+
+    entityPickerModal().findByText("Reviews").click();
```

Comment:
At this point `query` still has a join clause (the one with Products table), but UI displays it in draft state because it's incomplete after changing RHS table: [image](https://github.com/metabase/metabase/assets/6830683/303194b1-07b1-44cd-bca8-5c1940fc3e70)